packages/forma-36-react-components/src/components/Modal/ModalLauncher/ModalLauncher.tsx

```diff
@@ -0,0 +1,76 @@
+import ReactDOM from 'react-dom';
+
+type ModalLauncherComponentRendererProps<T = any> = {
+  isShown: boolean;
+  onClose: (result: T) => void;
+};
+
+type ModalLauncherOpenOptions = {
+  modalId?: string;
+  // ms before removing the component from the tree, defaults to 25ms.
+  delay?: number;
+};
+
+function open<T = any>(
+  componentRenderer: (
+    props: ModalLauncherComponentRendererProps<T>,
+  ) => JSX.Element,
+  options: ModalLauncherOpenOptions = {},
+): Promise<T> {
+  options = { delay: 300, ...options };
+
+  // Allow components to specify if they wish to reuse the modal container
+  const rootElId = `modals-root${options.modalId || Date.now()}`;
+  let rootDom: HTMLElement | null = null;
+
+  const getRoot = () => {
+    if (rootDom !== null) {
+      return rootDom;
+    }
+
+    // Reuse the container if we find it
+    rootDom = document.getElementById(rootElId);
+    if (rootDom !== null) {
+      return rootDom;
+    }
+
+    // Otherwise create it
+    rootDom = document.createElement('div');
+    rootDom.setAttribute('id', rootElId);
+    document.body.appendChild(rootDom);
+    return rootDom;
+  };
+
+  return new Promise(resolve => {
+    let currentConfig = { onClose, isShown: true };
+
+    function render({
+      onClose,
+      isShown,
+    }: ModalLauncherComponentRendererProps<T>) {
+      ReactDOM.render(componentRenderer({ onClose, isShown }), getRoot());
+    }
+
+    async function onClose(...args: T[]) {
```

Comment:
is it really `T[]`?

Reply:
i think yes because of the spread operator, it was complaining with just `T`, played around with it locally and autocompletion worked fine the way it is now

Reply:
I think it forces all args to be of type T with is not correct if we specify a concrete type and not `any`

Reply:
i'll look into it more

Reply:
restricted to 1 arg of type T as agreed